onChangeSelection update

Jmr3366 opened the pull request:
fixes #5079 

### Description of the Change
The is a large update `SelectModel`.

### Possible Drawbacks

This changes the behavior of `onChangeSelection` to only fire when a token selection change is made.
Previously, `onChangeSelection` would miss-fire multiple times on the same token.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5911)
<!-- Reviewable:end -->

Jmr3366 committed:
This is still a work in progress.

Jmr3366 commented:
At the moment, this will not fire on a selected token after you change maps.  I'd like to see if a developer could help determine if this is possible.

Jmr3366 committed:
This is still a work in progress.

Jmr3366 committed:
This is still a work in progress.

Jmr3366 committed:
Removed Zone changed check as it was not working.
Changed to using a MS (1500) delay check.

Jmr3366 committed:
Removed Zone changed check as it was not working.
Changed to using a MS (1500) delay check.

cwisniew commented:
> fixes #5079
> 
> ### Description of the Change
> The is a large update `SelectModel`.
> 
> ### Possible Drawbacks
> This changes the behavior of `onChangeSelection` to only fire when a token selection change is made. Previously, `onChangeSelection` would miss-fire multiple times on the same token.
> 
> This change is [<img alt="Reviewable" height="34" src="https://camo.githubusercontent.com/b672beca07dd3e7b25a0a923de4e20bb5ee5b858d23632bac38e8c42a5058d02/68747470733a2f2f72657669657761626c652e696f2f7265766965775f627574746f6e2e737667">](https://reviewable.io/reviews/RPTools/maptool/5911)

Have you tested this change to see if it still works the same way (minus duplicate events) when the user drag-selects the same token(s) that are already selected? Or if they use a macro to select the same token(s) that are already selected? i.e. You should get the same events before and after the change (without duplicates)

Jmr3366 commented:
> Have you tested this change to see if it still works the same way (minus duplicate events) when the user drag-selects the same token(s) that are already selected? Or if they use a macro to select the same token(s) that are already selected? i.e. You should get the same events before and after the change (without duplicates)

This [d2f9315](https://github.com/RPTools/maptool/pull/5911/commits/d2f93158c7bf27f72e0e251ef5ba2b9c29e43ca1) ,  does not work the way anyone would want but does in fact do what onChangeSelection is supposed to do, it will only fire when a selection changes,  You can't reselect the same token to trigger, a selection change of any kind is _required_ (no body would want this compared to what oCS does _today_).  The zone changed check will not achieve the desired end result and will not _work_.

This [d3b7f18](https://github.com/RPTools/maptool/pull/5911/commits/d3b7f18f24b8800a1e0845a0246db076e6f3bb14) , I believe is what would be expected for oCS.  It will only register a selection change when a selection change takes place, OR, after 2s of time have lapsed from a prior selectionChange.  I believe this is important to retain a non-bugged version of token re-selection that the current oCS allows.   A remaining issue I'd like to overcome is if you spam mouse-clicks on different tokens, _right now_  a 2s delay is being _enforced_ (not acceptable).  I'd like the 2s delay to only be enforced with the selection is the same.

You can drag box select, use `selectTokens`, add or remove tokens from a selection, all will trigger the event ONCE.  The only issue I see is to find a better way to handle the forced delay when you _want_ to change selection.  I'm going to work on the code to only enforce the 2s delay IF the selection is the same as before.

Jmr3366 committed:
Merge branch 'RPTools:develop' into jmr_develop

cwisniew commented:
If that is the case it  breaks backwards compatibility with macros. People may be depending on that behaviour. You would need to maintain that behaviour for the change to be included. Introduction of a timer to distinguish on change selection is also not really going to work either.

Jmr3366 commented:
> If that is the case it breaks backwards compatibility with macros. People may be depending on that behavior. You would need to maintain that behavior for the change to be included

I've just fixed the code for the 2s enforced delay ONLY if the selection is the same.

I'd like to say that _covers everything_

The timer isn't being used to distinguish selectionChange, is a _break_ to mitigate "reselection of the same thing".
(or maybe I'm missing your point).

edit
Now if you're saying script writers have code where they're wanting oCS to fire on the same token in under 2s, yes, that would fail.
edit x2
I've changed the time check to 300ms and it's working as _I'm expecting_

Jmr3366 committed:
Changed to using 250 MS delay check.
Allow trigger for re-selection of same token(s)

Jmr3366 commented:
The original state of SelectionModel could trigger selectedChanged() as little as 4 times but could trigger upwards to 10+.
This change minimizes the selectedChanged() trigger to _once_ when a _meaningful_ change takes place.  Meaningful should be any token selection change OR 250ms _after_ any prior token selection.

Jmr3366 committed:
Merge branch 'develop' into jmr_develop

Jmr3366 commented:
/denied
